woocommerce/woocommerce
# Add a generic function to determine remote logging eligibility

### Submission Review Guidelines:

-   I have followed the [WooCommerce Contributing Guidelines](https://github.com/woocommerce/woocommerce/blob/trunk/.github/CONTRIBUTING.md) and the [WordPress Coding Standards](https://make.wordpress.org/core/handbook/best-practices/coding-standards/).
-   I have checked to ensure there aren't other open [Pull Requests](https://github.com/woocommerce/woocommerce/pulls) for the same update/change.
-   I have reviewed my code for [security best practices](https://developer.wordpress.org/apis/security/).
-   Following the above guidelines will result in quick merges and clear and detailed feedback when appropriate.



### Changes proposed in this Pull Request:





Closes https://github.com/woocommerce/woocommerce/issues/49282

This PR adds the WC_Remote_Logger class and implements a function to determine if remote logging is allowed based on the following conditions. 

1. The remote logging feature flag is on. The remote logging feature flag is on. The feature is disabled by default and hidden from UI for now. To enable it, we need to set `woocommerce_feature_remote_logging_enabled` option to 'yes' 
2. The user has opted into tracking/logging.
3.  The current store WooCommerce version is the latest. This can ensure that we are not logging known issues that have been fixed in a later version. 
4. Check if the store is allowed to log in by checking the `woocommerce_remote_variant_assignment` option. We can use this option to gradually enable this feature for a percentage of users, starting with 10%.

These should cover the basic conditions to determine if remote logging is allowed, if you have any suggestions to improve this, feel free to comment. 🙏 



### How to test the changes in this Pull Request:



Using the [WooCommerce Testing Instructions Guide](https://github.com/woocommerce/woocommerce/wiki/Writing-high-quality-testing-instructions), include your detailed testing instructions:

This PR includes unit tests to test the `is_remote_logging_allowed` method, so please review the code and tests to ensure they are correct. The following manual tests can be used to test the implementation but feel free to do some of them if you are confident.

1. Create a fresh install of WooCommerce 
2. Opt-in to tracking during the Core Profiler
3. Install `Code Snippet` plugin
4. Add the following code snippet to your site:

```php
update_option( 'woocommerce_feature_remote_logging_enabled', 'yes' );

add_filter(
	'option_woocommerce_allow_tracking',
	function () {
		return 'yes';
	}
);

add_filter(
	'option_woocommerce_remote_variant_assignment',
	function () {
		return 5;
	}
);

set_transient( 'latest_woocommerce_version', '9.2.0', DAY_IN_SECONDS );


$remote_logger = new WC_Remote_Logger();
$logging_allowed = $remote_logger->is_remote_logging_allowed();
$log_message = $logging_allowed ? 'Remote logging allowed' : 'Remote logging not allowed';
error_log($log_message);

```

5. Check the debug log for the message "Remote logging allowed".
6. Update the code snippet to test the other conditions. It should log "Remote logging not allowed" if any of the below conditions are met:

- The remote error logging feature flag is OFF.
- The user has NOT opted into tracking/logging. 
- The current store WooCommerce version is not the latest.
- `woocommerce_remote_variant_assignment` option value is greater than 12.



### Changelog entry







<details>

<summary>Changelog Entry Details</summary>

#### Significance




#### Type




#### Message 

</details>

<details>

<summary>Changelog Entry Comment</summary>

#### Comment 

</details>
